File: src/main/java/dev/obscuria/tooltips/client/tooltip/filter/AllOfFilter.java
package dev.obscuria.tooltips.client.tooltip.filter;

import com.github.bsideup.jabel.Desugar;
import com.google.gson.JsonObject;
import net.minecraft.item.ItemStack;

import java.util.List;

@Desugar
public record AllOfFilter(List<ItemFilter> terms) implements ItemFilter {

    public static AllOfFilter fromJson(JsonObject json) {
        return new AllOfFilter(ItemFilter.parseTerms(json));
    }

    @Override
    public boolean test(ItemStack stack) {
        return terms.stream().allMatch(term -> term.test(stack));
    }
}

File: common/src/main/kotlin/dev/obscuria/tooltips/client/filter/AllOfFilter.kt
package dev.obscuria.tooltips.client.filter

import com.mojang.serialization.Codec
import com.mojang.serialization.codecs.RecordCodecBuilder
import net.minecraft.world.item.ItemStack

class AllOfFilter(
    val terms: List<IItemFilter>
) : IItemFilter {

    override fun codec(): Codec<AllOfFilter> = CODEC

    override fun test(stack: ItemStack): Boolean {
        return terms.all { it.test(stack) }
    }

    companion object {

        val CODEC: Codec<AllOfFilter> = RecordCodecBuilder.create { codec ->
            codec.group(
                IItemFilter.CODEC.listOf()
                    .fieldOf("terms")
                    .forGetter(AllOfFilter::terms)
            ).apply(codec, ::AllOfFilter)
        }
    }
}